### PR TITLE
Add metadata and plugins

### PR DESCRIPTION
**Pull request by zorkian**
This adds a very basic metadata (key, value) storage system, which can
be used by people who want to do things when accounts are set up, etc.
The plugins can be configured in the config_dir and apply to both the
API and FE servers.

### CODE REVIEW

**Comment by zorkian** [comment]
Even has tests on the metadata functions ;)

**Comment by zorkian** [comment]
@chet boom!

File: grouper/fe/util.py [hunk 1]
```diff
@@ -60,9 +60,9 @@ def get_current_user(self):
             raise InvalidUser()
 
         try:
-            user = self.session.query(User).filter_by(username=username).first()
-            if not user:
-                user = User(username=username).add(self.session)
+            user, created = User.get_or_create(self.session, username=username)
+            if created:
+                print "Created new user %s" % (username, )
```
Comment (jathanism) [inline grouper/fe/util.py]: `Should that `print still be in there? It seems out of place.

File: grouper/models.py [hunk 2]
```diff
@@ -118,6 +130,17 @@ def get_db_engine(url):
...
 
 
+Plugins = []
+
+
+def load_plugins(plugin_dir):
+    """Load plugins from a directory"""
+    global Plugins
+    if Plugins:
+        raise Exception("Plugins already loaded; can't load twice!")
```
Comment (jathanism) [inline grouper/models.py]: I feel like this should have its own exception. Something like:

```python
raise PluginsAlreadyLoaded("Plugins already loaded; can't load twice!")
```

That way you can catch that exception? Base `Exception` can be hard to isolate.

Comment (zorkian) [inline grouper/models.py]: Is it general good practice to make lots of named classes for
exceptions? I don't know what "best" is :)

--
Mark Smith mark@qq.is


On Sat, Feb 21, 2015, at 03:33 PM, Jathan McCollum wrote:
> In grouper/models.py[1]:


> > @@ -118,6 +130,17 @@ def get_db_engine(url):
>      return create_engine(url, pool_recycle=300)
>
>
> +Plugins = [] + + +def load_plugins(plugin_dir):
> +    """Load plugins from a directory"""
> +    global Plugins
> +    if Plugins:
> +        raise Exception("Plugins already loaded; can't load twice!")
>          I feel like this should have its own exception. Something
>          like:


> raise PluginsAlreadyLoaded("Plugins already loaded; can't load
> twice!") That way you can catch that exception? Base Exception can be
> hard to isolate.


> — Reply to this email directly or view it on GitHub[2].

>



Links:

  1. https://github.com/dropbox/grouper/pull/9#discussion_r25127047
  2. https://github.com/dropbox/grouper/pull/9/files#r25127047

Comment (jathanism) [inline grouper/models.py]: In short: yes. That way for the error-handling you can be very specific.

**Comment by zorkian** [comment]
Updated, thanks @jathanism !

**Pull request merged**